protocol/blockrelay/blockrelay.go

```diff
@@ -0,0 +1,274 @@
+	for hash := range blockHashes {
+		hash := hash
+		requestedBlocks.delete(&hash)
+	}
+}
+
+func requestBlocks(netAdapater *netadapter.NetAdapter, router *netadapter.Router, peer *peerpkg.Peer, msgChan <-chan wire.Message,
+	dag *blockdag.BlockDAG, invsQueue *[]*wire.MsgInvRelayBlock, requestQueue *[]*daghash.Hash,
+	requestQueueSet map[daghash.Hash]struct{}) (shouldStop bool, err error) {
+
+	var hashesToRequest []*daghash.Hash
+	if len(*requestQueue) > wire.MsgGetRelayBlocksHashes {
+		hashesToRequest, *requestQueue = (*requestQueue)[:wire.MsgGetRelayBlocksHashes],
+			(*requestQueue)[wire.MsgGetRelayBlocksHashes:]
+	} else {
+		hashesToRequest, *requestQueue = *requestQueue, nil
+	}
+
+	pendingBlocks := map[daghash.Hash]struct{}{}
+	for _, hash := range hashesToRequest {
+		delete(requestQueueSet, *hash)
+		pendingBlocks[*hash] = struct{}{}
+		exists := requestedBlocks.addIfExists(hash)
+		if exists {
+			return false, nil
+		}
+	}
+
+	getRelayBlocksMsg := wire.NewMsgGetRelayBlocks(hashesToRequest)
+	router.WriteOutgoingMessage(getRelayBlocksMsg)
+
+	for len(pendingBlocks) > 0 {
+		msgBlock, shouldStop, err := readMsgBlock(msgChan, invsQueue)
+		if err != nil {
+			return false, err
+		}
+		if shouldStop {
+			return true, nil
+		}
+
+		block := util.NewBlock(msgBlock)
+		blockHash := block.Hash()
+		if _, ok := pendingBlocks[*blockHash]; !ok {
+			return false, errors.Errorf("got unrequested block %s", block.Hash())
+		}
+		delete(pendingBlocks, *blockHash)
+		requestedBlocks.delete(blockHash)
+
+		shouldStop, err = processAndRelayBlock(netAdapater, router, peer, dag, requestQueue, requestQueueSet, block)
+		if err != nil {
+			return false, err
+		}
+		if shouldStop {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func processAndRelayBlock(netAdapter *netadapter.NetAdapter, router *netadapter.Router, peer *peerpkg.Peer,
+	dag *blockdag.BlockDAG, requestQueue *[]*daghash.Hash, requestQueueSet map[daghash.Hash]struct{},
+	block *util.Block) (shouldStop bool, err error) {
+
+	blockHash := block.Hash()
+	isOrphan, isDelayed, err := dag.ProcessBlock(block, blockdag.BFNone)
+	if err != nil {
+		// When the error is a rule error, it means the block was simply
+		// rejected as opposed to something actually going wrong, so log
+		// it as such. Otherwise, something really did go wrong, so log
+		// it as an actual error.
+		if !errors.As(err, &blockdag.RuleError{}) {
+			panic(errors.Wrapf(err, "failed to process block %s",
+				blockHash))
+		}
+		log.Infof("Rejected block %s from %s: %s", blockHash,
+			peer, err)
+
+		return false, errors.Wrap(err, "got invalid block: %s")
+	}
+
+	if isDelayed {
+		return false, nil
+	}
+
+	if isOrphan {
+		blueScore, err := block.BlueScore()
+		if err != nil {
+			log.Errorf("Received an orphan block %s with malformed blue score from %s. Disconnecting...",
+				blockHash, peer)
+			return false, errors.Errorf("Received an orphan block %s with malformed blue score", blockHash)
```

Comment:
Don't we want to return shouldStop=true if disconnecting?

Reply:
No, shouldStop=true only if the channel is closed